secure/storage/src/storage.rs

```diff
@@ -25,3 +25,34 @@ pub enum Storage {
     NamespacedStorage(NamespacedStorage),
     OnDiskStorage(OnDiskStorage),
 }
+
+impl Storage {
+    pub fn into_cached(self) -> CachedStorage {
+        match self {
+            Storage::GitHubStorage(storage) => CachedStorage::GitHubStorage(storage.into_cached()),
+            Storage::VaultStorage(storage) => CachedStorage::VaultStorage(storage.into_cached()),
+            Storage::InMemoryStorage(storage) => {
+                CachedStorage::InMemoryStorage(storage.into_cached())
+            }
+            Storage::NamespacedStorage(storage) => {
+                CachedStorage::NamespacedStorage(storage.into_cached())
+            }
+            Storage::OnDiskStorage(storage) => CachedStorage::OnDiskStorage(storage.into_cached()),
+        }
+    }
+}
+
+#[enum_dispatch(KVStorage, CryptoStorage)]
+pub enum CachedStorage {
+    GitHubStorage(CachedKVStorage<GitHubStorage>),
+    VaultStorage(CachedKVStorage<VaultStorage>),
+    InMemoryStorage(CachedKVStorage<InMemoryStorage>),
+    NamespacedStorage(CachedKVStorage<NamespacedStorage>),
+    OnDiskStorage(CachedKVStorage<OnDiskStorage>),
+}
+
+#[enum_dispatch(KVStorage, CryptoStorage)]
+pub enum StorageVariant {
+    Storage(Storage),
+    CachedStorage(CachedStorage),
+}
```

Comment:
Considering that we are only caching the KV side, could we do something like how the NamespacedStorage works? That way we don't end up with this additional layer of indirection? I would also probably consider just reusing the the concept with how that works rather than introducing "into_cached", unless we want to add a new feature "into_namespaced" which would be kinda hard.

Other than that, this looks good!

As future work, we should consider that each Storage interface is just a client to a backend and make them not Send + Sync and then we can get rid of the locks on the cache and instead make them within the storage itself -- if necessary.

Reply:
I see, so are we thinking about replacing the original `Storage` enum with a struct with a Boxed storage inside, like
```
pub enum Storage {
    storage: Box<dyn KVStorage + CryptoStorage>
}
```
so that `storage` can be either a normal `GitHubStorage`, `VaultStorage`, etc., or their cached equivalent?
For the `IntoCached` trait, I added it only for ergonomics, and can be removed without any problems, in which case we would just call `CachedStorage::new` in its place :) Please let me know if that's what you have in mind. Thanks!!

Reply:
Yeah, let's keep CachedStorage to be similar to NamespacedStorage and then we should be good to go with this PR, right?

Reply:
Okay, I renamed a few things and used `Box`. Please let me know if they look okay, thanks!

Reply:
Actually should we also cache the public and private keys? I'm not sure if they'll be modified by other processes, which would result in a stale cache..

Reply:
yeah, those are modifiable by other processes, so we do not want to cache them. That's why I had wondered if we should have just made a cache specific to LSR by adding the layer to PersistentSafetyStorage. But this is good too!

Reply:
Hmm if that's the case then I'd have to make sure the CachedStorage doesn't cache the public and private keys, cuz right now they're caching them...

Reply:
oh, that might explain why I cannot run the perf test... there's no logging enabled right now for safety rules, I'm trying to fix that in another PR :)

Reply:
Haha, okay I just changed the cache so that it doesn't cache the public/private keys.. Please let me know if it works now, thanks!